cmake/f3dPlugin.cmake

```diff
@@ -0,0 +1,211 @@
+    set(F3D_READER_HAS_SCENE_READER 1)
+  else()
+    set(F3D_READER_HAS_SCENE_READER 0)
+  endif()
+
+  if(F3D_READER_VTK_READER)
+    set(F3D_READER_HAS_GEOMETRY_READER 1)
+  else()
+    set(F3D_READER_HAS_GEOMETRY_READER 0)
+  endif()
+
+  configure_file("${CMAKE_SOURCE_DIR}/cmake/readerBoilerPlate.h.in"
+    "${CMAKE_CURRENT_BINARY_DIR}/reader_${F3D_READER_NAME}.h")
+endmacro()
+
+#[==[
+@brief Build the plugin.
+
+~~~
+f3d_plugin_build(
+  NAME                   <string>
+  [DESCRIPTION           <string>]
+  [VERSION               <string>]
+  [VTK_MODULES           <string>...]
+  [ADDITIONAL_RPATHS     <path>...]
+  [FORCE_STATIC])
+~~~
+
+Declare the plugin. Must be called after ``f3d_plugin_declare_reader``.
+
+The `NAME` argument is required. The arguments are as follows:
+
+  * `NAME`: (Required) The name of the plugin.
+  * `DESCRIPTION`: The description of the plugin.
+  * `VERSION`: The version of the plugin.
+  * `VTK_MODULES`: The list of VTK modules used by the plugin to link with.
+  * `ADDITIONAL_RPATHS`: The list of additional RPATH for the installed binaries on Unix. VTK path is added automatically.
+  * `FORCE_STATIC`: If specified, the plugin is built as a static library and embedded into libf3d.
+#]==]
+
+macro(f3d_plugin_build)
+  cmake_parse_arguments(F3D_PLUGIN "FORCE_STATIC" "NAME;DESCRIPTION;VERSION" "VTK_MODULES;ADDITIONAL_RPATHS" ${ARGN})
+
+  if(F3D_PLUGIN_FORCE_STATIC OR F3D_PLUGINS_STATIC_BUILD)
+    set(F3D_PLUGIN_TYPE "STATIC")
+    set(F3D_PLUGIN_IS_STATIC ON)
+    set_property(GLOBAL APPEND PROPERTY F3D_STATIC_PLUGINS ${F3D_PLUGIN_NAME})
+  else()
+    set(F3D_PLUGIN_TYPE "MODULE")
+    set(F3D_PLUGIN_IS_STATIC OFF)
+
+    get_target_property(target_type VTK::CommonCore TYPE)
+    if (target_type STREQUAL STATIC_LIBRARY)
+      message(FATAL_ERROR "\"${F3D_PLUGIN_NAME}\" plugin is built as a shared module, but VTK is built as a static library. Please flag the plugin as FORCE_STATIC, enable option F3D_PLUGINS_STATIC_BUILD, or rebuild VTK as a shared library.")
+    endif ()
+  endif()
+
+  # if libf3d is built as a static library, VTK extension static libraries
+  # must be exported as well for proper linkage
+  set(export_name "")
+  if(NOT BUILD_SHARED_LIBS AND F3D_PLUGIN_IS_STATIC)
+    set(export_name "f3dTargets")
+  endif()
+
+  find_package(VTK 9.0 REQUIRED COMPONENTS ${F3D_PLUGIN_VTK_MODULES})
+
+  vtk_module_find_modules(vtk_module_files "${CMAKE_CURRENT_SOURCE_DIR}")
+
+  if(NOT ${vtk_module_files} STREQUAL "")
+    vtk_module_scan(
+      MODULE_FILES            ${vtk_module_files}
+      PROVIDES_MODULES        modules
+      WANT_BY_DEFAULT         ON
+      HIDE_MODULES_FROM_CACHE ON
+      ENABLE_TESTS            ${BUILD_TESTING})
+
+    vtk_module_build(
+      MODULES ${modules}
+      INSTALL_EXPORT ${export_name}
+      INSTALL_HEADERS OFF
+      HEADERS_COMPONENT vtkext
+      TARGETS_COMPONENT vtkext
+      PACKAGE "f3d-plugin-${F3D_PLUGIN_NAME}")
+  else()
+    set(modules "")
```

Comment:
can this happen ?  the module will at least depends on something in VTK.

Reply:
Yes, in the case of "core" and "exodus", these plugins have no VTK modules.

Reply:
haaa, this list only VTKExtensions modules, ok.